File: app/src/main/java/com/dmi/perfectreader/layout/liner/Liner.java
package com.dmi.perfectreader.layout.liner;

import java.util.List;
import java.util.Locale;

public interface Liner {
    List<Line> makeLines(MeasuredText measuredText, Config config);

    interface Line {
        float left();
        float width();
        boolean hasHyphenAfter();
        boolean isLast();
        List<Token> tokens();
    }

    interface Token {
        boolean isSpace();
        int beginIndex();
        int endIndex();
    }

    interface MeasuredText {
        CharSequence plainText();
        Locale locale();
        float widthOf(int index);
        float widthOf(int beginIndex, int endIndex);
        float hyphenWidthAfter(int index);
    }

    interface Config {
        float firstLineIndent();
        float maxWidth();
        float leftHangFactor(char ch);
        float rightHangFactor(char ch);
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/layout/liner/Liner.kt
package com.dmi.perfectreader.layout.liner

import java.util.*

interface Liner {
    fun makeLines(measuredText: MeasuredText, config: Config): List<Line>

    class Line {
        var left = 0F
        var width = 0F
        var hasHyphenAfter = false
        var tokens = ArrayList<Token>()
        val right: Float
            get() = left + width
    }

    class Token {
        var isSpace = false
        var beginIndex = 0
        var endIndex = 0
    }

    interface MeasuredText {
        val plainText: CharSequence
        val locale: Locale
        fun widthOf(index: Int): Float
        fun widthOf(beginIndex: Int, endIndex: Int): Float
        fun hyphenWidthAfter(index: Int): Float
    }

    interface Config {
        val firstLineIndent: Float
        val maxWidth: Float
        fun leftHangFactor(ch: Char): Float
        fun rightHangFactor(ch: Char): Float
    }
}
